ambry-network/src/main/java/com/github/ambry/network/AdaptiveLifoCoDelNetworkRequestQueue.java

```diff
@@ -62,28 +61,34 @@ public class AdaptiveLifoCoDelNetworkRequestQueue implements NetworkRequestQueue
   private volatile boolean emptyRequestReceived = false;
 
   /**
-   *  @param lifoThreshold the fraction of capacity used at which to switch the queue from FIFO to LIFO mode.
+   * @param lifoThreshold      the fraction of capacity used at which to switch the queue from FIFO to LIFO mode.
    * @param coDelTargetDelayMs the target delay in ms to use for the controlled delay algorithm.
-   * @param coDelIntervalMs the target interval in ms to use for the controlled delay algorithm.
-   * @param time {@link Time} instance.
+   * @param coDelIntervalMs    the target interval in ms to use for the controlled delay algorithm.
+   * @param time               {@link Time} instance.
+   * @param capacity           the capacity of the queue
    */
-  AdaptiveLifoCoDelNetworkRequestQueue(int lifoThreshold, int coDelTargetDelayMs, int coDelIntervalMs, Time time) {
+  AdaptiveLifoCoDelNetworkRequestQueue(int lifoThreshold, int coDelTargetDelayMs, int coDelIntervalMs, Time time,
+      int capacity) {
     this.coDelTargetDelayMs = coDelTargetDelayMs;
     this.coDelIntervalMs = coDelIntervalMs;
     this.lifoThreshold = lifoThreshold;
     this.time = time;
     intervalTimeMs = time.milliseconds();
+    deque = new LinkedBlockingDeque<>(capacity);
   }
 
   @Override
-  public void offer(NetworkRequest request) throws InterruptedException {
-    deque.offer(request);
-    if (request.equals(EmptyRequest.getInstance())) {
-      // This is a internal generated request used to signal shutdown. Add it to dropped requests queue as well so that
-      // request-dropper thread shuts down gracefully.
-      droppedRequestsQueue.offer(request);
-      emptyRequestReceived = true;
+  public boolean offer(NetworkRequest request) throws InterruptedException {
+    if (deque.offer(request)) {
+      if (request.equals(EmptyRequest.getInstance())) {
```

Comment:
When we receive the shutdown/EmptyRequest, actually we don't need handle the existing queue. We can drop all the requests in queue.
Will it make it simple?

Reply:
@JingQianCloud You are right. If the queue is full, the EmptyRequest would be dropped. For this actually, I added a shared variable to shut down the request handler thread here https://github.com/linkedin/ambry/blob/9c685c5389ff6c07523a0cba89c16781b882a4b7/ambry-protocol/src/main/java/com/github/ambry/protocol/RequestHandler.java#L66.

Yeah, we could also drop existing requests to populate the EmptyRequest. But I think we may need to add synchronization then. I thought this may be easier

Reply:
Got you.